// @azure/monitor-opentelemetry-exporter/dist-esm/src/utils/spanUtils.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readableSpanToEnvelope = void 0;
var url_1 = require("url");
var core_1 = require("@opentelemetry/core");
var api_1 = require("@opentelemetry/api");
var httpAttributes_1 = require("./constants/span/httpAttributes");
var applicationinsights_1 = require("./constants/applicationinsights");
var grpcAttributes_1 = require("./constants/span/grpcAttributes");
var breezeUtils_1 = require("./breezeUtils");
var platform_1 = require("../platform");
var dbAttributes_1 = require("./constants/span/dbAttributes");
var eventhub_1 = require("./eventhub");
var azAttributes_1 = require("./constants/span/azAttributes");
function createTagsFromSpan(span) {
    var context = platform_1.getInstance();
    var tags = __assign({}, context.tags);
    tags[applicationinsights_1.AI_OPERATION_ID] = span.spanContext.traceId;
    if (span.parentSpanId) {
        tags[applicationinsights_1.AI_OPERATION_PARENT_ID] = span.parentSpanId;
    }
    // @todo: is this for RequestData only?
    if ((span.kind === api_1.SpanKind.SERVER || span.kind === api_1.SpanKind.CONSUMER) &&
        span.attributes[grpcAttributes_1.GRPC_METHOD]) {
        tags[applicationinsights_1.AI_OPERATION_NAME] = String(span.attributes[grpcAttributes_1.GRPC_METHOD]);
    }
    if ((span.kind === api_1.SpanKind.SERVER || span.kind === api_1.SpanKind.CONSUMER) &&
        span.attributes[httpAttributes_1.HTTP_METHOD] &&
        span.attributes[httpAttributes_1.HTTP_ROUTE]) {
        tags[applicationinsights_1.AI_OPERATION_NAME] = span.attributes[httpAttributes_1.HTTP_METHOD] + " " + span.attributes[httpAttributes_1.HTTP_ROUTE];
    }
    return tags;
}
function createPropertiesFromSpan(span) {
    var properties = {};
    var measurements = {};
    for (var _i = 0, _a = Object.keys(span.attributes); _i < _a.length; _i++) {
        var key = _a[_i];
        if (key === grpcAttributes_1.GRPC_ERROR_MESSAGE ||
            key === grpcAttributes_1.GRPC_ERROR_NAME ||
            !(key.startsWith("http.") || key.startsWith("grpc.") || key.startsWith("db."))) {
            properties[key] = span.attributes[key];
        }
    }
    var links = span.links.map(function (link) { return ({
        operation_Id: link.context.traceId,
        id: link.context.spanId
    }); });
    if (links.length > 0) {
        properties[applicationinsights_1.MS_LINKS] = JSON.stringify(links);
    }
    return [properties, measurements];
}
function createDependencyData(span) {
    var data = {
        name: span.name,
        id: "|" + span.spanContext.traceId + "." + span.spanContext.spanId + ".",
        success: span.status.code === api_1.CanonicalCode.OK,
        resultCode: String(span.status.code),
        target: span.attributes[httpAttributes_1.HTTP_URL],
        type: "Dependency",
        duration: breezeUtils_1.msToTimeSpan(core_1.hrTimeToMilliseconds(span.duration)),
        version: 1
    };
    if (span.attributes[httpAttributes_1.HTTP_STATUS_CODE]) {
        data.type = "HTTP";
        data.resultCode = String(span.attributes[httpAttributes_1.HTTP_STATUS_CODE]);
    }
    if (span.attributes[grpcAttributes_1.GRPC_STATUS_CODE] !== undefined) {
        data.type = "GRPC";
        data.resultCode = String(span.attributes[grpcAttributes_1.GRPC_STATUS_CODE]);
    }
    if (span.attributes[grpcAttributes_1.GRPC_METHOD]) {
        data.target = String(span.attributes[grpcAttributes_1.GRPC_METHOD]);
        data.data = String(span.attributes[grpcAttributes_1.GRPC_METHOD]);
    }
    if (span.attributes[httpAttributes_1.HTTP_URL]) {
        var url = new url_1.URL(span.attributes[httpAttributes_1.HTTP_URL]);
        data.target = url.hostname;
        data.data = url.href;
        if (span.attributes[httpAttributes_1.HTTP_METHOD]) {
            data.name = span.attributes[httpAttributes_1.HTTP_METHOD] + " " + url.pathname;
        }
    }
    if (span.attributes[dbAttributes_1.DB_STATEMENT]) {
        data.name = String(span.attributes[dbAttributes_1.DB_STATEMENT]);
        data.data = String(span.attributes[dbAttributes_1.DB_STATEMENT]);
        if (span.attributes[dbAttributes_1.DB_TYPE]) {
            data.type = String(span.attributes[dbAttributes_1.DB_TYPE]);
        }
        else {
            data.type = "DB";
        }
        if (span.attributes[dbAttributes_1.DB_INSTANCE]) {
            data.target = String(span.attributes[dbAttributes_1.DB_INSTANCE]);
        }
    }
    return data;
}
function createRequestData(span) {
    var data = {
        name: span.name,
        id: "|" + span.spanContext.traceId + "." + span.spanContext.spanId + ".",
        success: span.status.code === api_1.CanonicalCode.OK,
        responseCode: String(span.status.code),
        duration: breezeUtils_1.msToTimeSpan(core_1.hrTimeToMilliseconds(span.duration)),
        version: 1,
        source: undefined
    };
    if (span.attributes[httpAttributes_1.HTTP_METHOD]) {
        data.name = span.attributes[httpAttributes_1.HTTP_METHOD];
        if (span.attributes[httpAttributes_1.HTTP_STATUS_CODE]) {
            data.responseCode = String(span.attributes[httpAttributes_1.HTTP_STATUS_CODE]);
        }
        if (span.attributes[httpAttributes_1.HTTP_URL]) {
            data.url = span.attributes[httpAttributes_1.HTTP_URL];
        }
        if (span.attributes[httpAttributes_1.HTTP_ROUTE]) {
            data.name = span.attributes[httpAttributes_1.HTTP_METHOD] + " " + span.attributes[httpAttributes_1.HTTP_ROUTE];
        }
        else if (span.attributes[httpAttributes_1.HTTP_URL]) {
            var url = new url_1.URL(span.attributes[httpAttributes_1.HTTP_URL]);
            data.name = span.attributes[httpAttributes_1.HTTP_METHOD] + " " + url.pathname;
        }
    }
    if (span.attributes[grpcAttributes_1.GRPC_STATUS_CODE]) {
        data.responseCode = String(span.attributes[grpcAttributes_1.GRPC_STATUS_CODE]);
    }
    if (span.attributes[grpcAttributes_1.GRPC_METHOD]) {
        data.url = String(span.attributes[grpcAttributes_1.GRPC_METHOD]);
    }
    return data;
}
function createInProcData(span) {
    var data = createDependencyData(span);
    data.type = applicationinsights_1.INPROC;
    data.success = true;
    return data;
}
function readableSpanToEnvelope(span, ikey, logger) {
    var name;
    var baseType;
    var sampleRate = 100;
    var baseData;
    var time = new Date(core_1.hrTimeToMilliseconds(span.startTime));
    var instrumentationKey = ikey;
    var tags = createTagsFromSpan(span);
    var _a = createPropertiesFromSpan(span), properties = _a[0], measurements = _a[1];
    switch (span.kind) {
        case api_1.SpanKind.CLIENT:
        case api_1.SpanKind.PRODUCER:
            name = "Microsoft.ApplicationInsights.RemoteDependency";
            baseType = "RemoteDependencyData";
            baseData = createDependencyData(span);
            break;
        case api_1.SpanKind.SERVER:
        case api_1.SpanKind.CONSUMER:
            name = "Microsoft.ApplicationInsights.Request";
            baseType = "RequestData";
            baseData = createRequestData(span);
            break;
        case api_1.SpanKind.INTERNAL:
            baseType = "RemoteDependencyData";
            name = "Microsoft.ApplicationInsights.RemoteDependency";
            baseData = createInProcData(span);
            break;
        default:
            // never
            if (logger) {
                logger.error("Unsupported span kind " + span.kind);
            }
            throw new Error("Unsupported span kind " + span.kind);
    }
    if (span.attributes[azAttributes_1.AzNamespace] === azAttributes_1.MicrosoftEventHub) {
        eventhub_1.parseEventHubSpan(span, baseData);
    }
    else if (span.attributes[azAttributes_1.AzNamespace]) {
        switch (span.kind) {
            case api_1.SpanKind.INTERNAL:
                baseData.type = applicationinsights_1.INPROC + " | " + span.attributes[azAttributes_1.AzNamespace];
                break;
            default: // no op
        }
    }
    return {
        name: name,
        sampleRate: sampleRate,
        time: time,
        instrumentationKey: instrumentationKey,
        tags: tags,
        version: 1,
        data: {
            baseType: baseType,
            baseData: __assign(__assign({}, baseData), { properties: properties,
                measurements: measurements })
        }
    };
}
exports.readableSpanToEnvelope = readableSpanToEnvelope;
