// @azure/monitor-opentelemetry-exporter/dist-esm/src/export/exporter.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureMonitorBaseExporter = void 0;
var core_1 = require("@opentelemetry/core");
var connectionStringParser_1 = require("../utils/connectionStringParser");
var platform_1 = require("../platform");
var config_1 = require("../config");
var breezeUtils_1 = require("../utils/breezeUtils");
var Constants_1 = require("../Declarations/Constants");
var AzureMonitorBaseExporter = /** @class */ (function () {
    function AzureMonitorBaseExporter(_options) {
        if (_options === void 0) { _options = {}; }
        var connectionString = _options.connectionString || process.env[Constants_1.ENV_CONNECTION_STRING];
        var instrumentationKey = _options.instrumentationKey || process.env[Constants_1.ENV_INSTRUMENTATION_KEY] || "";
        this._logger = _options.logger || new core_1.ConsoleLogger(core_1.LogLevel.ERROR);
        this._options = __assign(__assign(__assign({}, config_1.DEFAULT_EXPORTER_CONFIG), _options), { instrumentationKey: instrumentationKey });
        if (connectionString) {
            var parsedConnectionString = connectionStringParser_1.ConnectionStringParser.parse(connectionString, this._logger);
            this._options = __assign(__assign({}, config_1.DEFAULT_EXPORTER_CONFIG), { 
                // Overwrite options with connection string results, if any
                instrumentationKey: parsedConnectionString.instrumentationkey || instrumentationKey, endpointUrl: parsedConnectionString.ingestionendpoint || _options.endpointUrl });
        }
        // Instrumentation key is required
        if (!this._options.instrumentationKey) {
            var message = "No instrumentation key or connection string was provided to the Azure Monitor Exporter";
            this._logger.error(message);
            throw new Error(message);
        }
        this._telemetryProcessors = [];
        this._sender = new platform_1.HttpSender();
        this._persister = new platform_1.FileSystemPersist(this._options);
        this._retryTimer = null;
        this._logger.debug("AzureMonitorTraceExporter was successfully setup");
    }
    AzureMonitorBaseExporter.prototype._persist = function (envelopes) {
        return __awaiter(this, void 0, void 0, function () {
            var success, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._persister.push(envelopes)];
                    case 1:
                        success = _a.sent();
                        return [2 /*return*/, success ? core_1.ExportResult.FAILED_RETRYABLE : core_1.ExportResult.FAILED_NOT_RETRYABLE];
                    case 2:
                        e_1 = _a.sent();
                        return [2 /*return*/, core_1.ExportResult.FAILED_NOT_RETRYABLE];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    AzureMonitorBaseExporter.prototype.exportEnvelopes = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var envelopes, _a, result, statusCode, breezeResponse, filteredEnvelopes, senderErr_1;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        envelopes = this._applyTelemetryProcessors(payload);
                        this._logger.info("Exporting " + envelopes.length + " envelope(s)");
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 10, , 11]);
                        return [4 /*yield*/, this._sender.send(envelopes)];
                    case 2:
                        _a = _b.sent(), result = _a.result, statusCode = _a.statusCode;
                        if (!(statusCode === 200)) return [3 /*break*/, 3];
                        // Success -- @todo: start retry timer
                        if (!this._retryTimer) {
                            this._retryTimer = setTimeout(function () {
                                _this._retryTimer = null;
                                _this._sendFirstPersistedFile();
                            }, this._options.batchSendRetryIntervalMs);
                            this._retryTimer.unref();
                        }
                        return [2 /*return*/, core_1.ExportResult.SUCCESS];
                    case 3:
                        if (!(statusCode && breezeUtils_1.isRetriable(statusCode))) return [3 /*break*/, 8];
                        if (!result) return [3 /*break*/, 5];
                        this._logger.info(result);
                        breezeResponse = JSON.parse(result);
                        filteredEnvelopes = breezeResponse.errors.reduce(function (acc, v) { return __spreadArrays(acc, [envelopes[v.index]]); }, []);
                        return [4 /*yield*/, this._persist(filteredEnvelopes)];
                    case 4: 
                    // calls resultCallback(ExportResult) based on result of persister.push
                    return [2 /*return*/, _b.sent()];
                    case 5: return [4 /*yield*/, this._persist(envelopes)];
                    case 6: 
                    // calls resultCallback(ExportResult) based on result of persister.push
                    return [2 /*return*/, _b.sent()];
                    case 7: return [3 /*break*/, 9];
                    case 8: 
                    // Failed -- not retriable
                    return [2 /*return*/, core_1.ExportResult.FAILED_NOT_RETRYABLE];
                    case 9: return [3 /*break*/, 11];
                    case 10:
                        senderErr_1 = _b.sent();
                        // Request failed -- always retry
                        this._logger.error(senderErr_1.message);
                        return [2 /*return*/, this._persist(envelopes)];
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    AzureMonitorBaseExporter.prototype.addTelemetryProcessor = function (processor) {
        this._telemetryProcessors.push(processor);
    };
    AzureMonitorBaseExporter.prototype.clearTelemetryProcessors = function () {
        this._telemetryProcessors = [];
    };
    AzureMonitorBaseExporter.prototype.shutdown = function () {
        this._sender.shutdown();
    };
    AzureMonitorBaseExporter.prototype._applyTelemetryProcessors = function (envelopes) {
        var filteredEnvelopes = [];
        for (var _i = 0, envelopes_1 = envelopes; _i < envelopes_1.length; _i++) {
            var envelope = envelopes_1[_i];
            var accepted = true;
            for (var _a = 0, _b = this._telemetryProcessors; _a < _b.length; _a++) {
                var processor = _b[_a];
                // Don't use CPU cycles if item is already rejected
                if (accepted && processor(envelope) === false) {
                    accepted = false;
                }
            }
            if (accepted) {
                filteredEnvelopes.push(envelope);
            }
        }
        return filteredEnvelopes;
    };
    AzureMonitorBaseExporter.prototype._sendFirstPersistedFile = function () {
        return __awaiter(this, void 0, void 0, function () {
            var envelopes, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._persister.shift()];
                    case 1:
                        envelopes = (_a.sent());
                        this._sender.send(envelopes);
                        return [3 /*break*/, 3];
                    case 2:
                        err_1 = _a.sent();
                        this._logger.warn("Failed to fetch persisted file", err_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return AzureMonitorBaseExporter;
}());
exports.AzureMonitorBaseExporter = AzureMonitorBaseExporter;
