// @azure/monitor-opentelemetry-exporter/dist-esm/src/generated/applicationInsightsClient.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationInsightsClient = void 0;
var coreHttp = __importStar(require("@azure/core-http"));
var Parameters = __importStar(require("./models/parameters"));
var Mappers = __importStar(require("./models/mappers"));
var applicationInsightsClientContext_1 = require("./applicationInsightsClientContext");
var ApplicationInsightsClient = /** @class */ (function (_super) {
    __extends(ApplicationInsightsClient, _super);
    /**
     * Initializes a new instance of the ApplicationInsightsClient class.
     * @param options The parameter options
     */
    function ApplicationInsightsClient(options) {
        return _super.call(this, options) || this;
    }
    /**
     * This operation sends a sequence of telemetry events that will be monitored by Azure Monitor.
     * @param body The list of telemetry events to track.
     * @param options The options parameters.
     */
    ApplicationInsightsClient.prototype.track = function (body, options) {
        var operationOptions = coreHttp.operationOptionsToRequestOptionsBase(options || {});
        return this.sendOperationRequest({ body: body, options: operationOptions }, trackOperationSpec);
    };
    return ApplicationInsightsClient;
}(applicationInsightsClientContext_1.ApplicationInsightsClientContext));
exports.ApplicationInsightsClient = ApplicationInsightsClient;
// Operation Specifications
var serializer = new coreHttp.Serializer(Mappers, /* isXml */ false);
var trackOperationSpec = {
    path: "/track",
    httpMethod: "POST",
    responses: {
        200: {
            bodyMapper: Mappers.TrackResponse
        },
        206: {
            bodyMapper: Mappers.TrackResponse
        },
        400: {
            bodyMapper: Mappers.TrackResponse
        },
        402: {
            bodyMapper: Mappers.TrackResponse
        },
        429: {
            bodyMapper: Mappers.TrackResponse
        },
        500: {
            bodyMapper: Mappers.TrackResponse
        },
        503: {
            bodyMapper: Mappers.TrackResponse
        }
    },
    requestBody: Parameters.body,
    urlParameters: [Parameters.host],
    headerParameters: [Parameters.contentType, Parameters.accept],
    mediaType: "json",
    serializer: serializer
};
