// @azure/monitor-opentelemetry-exporter/dist-esm/src/utils/constants/span/dbAttributes.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PEER_SERVICE = exports.PEER_IPV6 = exports.PEER_IPV4 = exports.PEER_PORT = exports.PEER_HOSTNAME = exports.PEER_ADDRESS = exports.DB_STATEMENT = exports.DB_INSTANCE = exports.DB_TYPE = void 0;
var conventions = __importStar(require("@opentelemetry/semantic-conventions"));
exports.DB_TYPE = conventions.DatabaseAttribute.DB_TYPE;
exports.DB_INSTANCE = conventions.DatabaseAttribute.DB_INSTANCE;
exports.DB_STATEMENT = conventions.DatabaseAttribute.DB_STATEMENT;
exports.PEER_ADDRESS = conventions.GeneralAttribute.NET_PEER_ADDRESS;
exports.PEER_HOSTNAME = conventions.GeneralAttribute.NET_PEER_HOSTNAME;
// Optionals
exports.PEER_PORT = conventions.GeneralAttribute.NET_PEER_PORT;
exports.PEER_IPV4 = conventions.GeneralAttribute.NET_PEER_IPV4;
exports.PEER_IPV6 = conventions.GeneralAttribute.NET_PEER_IPV6;
exports.PEER_SERVICE = conventions.GeneralAttribute.NET_PEER_SERVICE;
